### PR TITLE
Handle Content-Type from url that point to binary instead of assuming its HTTPDirectory;

### PR DESCRIPTION
**Pull request by bigretromike**
## Description

When checking if directory should be cached, also check if the item that is being cached is not empty, if so skip caching.

## Motivation and context

While trying to play http source, Kodi try to scan it for external subtitles (correct) but even there is non it try to cache url, breaking later attempt to play same url (video plugin).
This PR should resolve issue with:
https://github.com/xbmc/xbmc/issues/21495
and probably is related to this:
https://github.com/xbmc/xbmc/issues/20151

## How has this been tested?

Was tested with default installation

## What is the effect on users?

This shouldn't impact any user

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by enen92** [comment]
Thanks for the contribution. I think caching a directory is a valid use case regardless of it being empty or not (it just speeds up access). By the sound of your description Kodi is doing something wrong but certainly not in the place you changed. Supposing this is some sort of remote FS (http directory?) the server likely indicates some error code that we probably could use to avoid caching. Another possibility might be to set a property in the listitem to make Kodi avoid caching.

**Comment by bigretromike** [comment]
> 

here is the issue this PR resolve: https://github.com/xbmc/xbmc/issues/21495 
The issue is that Kodi try to search Subtitles for the url=`http://<ip>:<port>/Stream/<id>/False/file.mkv` and even when it does not find anything it cache the directory as empty, then play file. The second time kodi try to play file it Parse `url`, find it in cache, return empty item, and throw error and cannot play the url without exiting Addon, and reentering (which clears the cache). 
I tried to pin-point the issue but somehow I dont see any other code that I could tweak to resolve it, and not caching empty object was a good idea (probably not the best)

**Comment by bigretromike** [comment]
> Thanks for the contribution. I think caching a directory is a valid use case regardless of it being empty or not (it just speeds up access). By the sound of your description Kodi is doing something wrong but certainly not in the place you changed. Supposing this is some sort of remote FS (http directory?) the server likely indicates some error code that we probably could use to avoid caching. Another possibility might be to set a property in the listitem to make Kodi avoid caching.

I tried few thing that imo should work but nothing fix this:
```
item = xbmcgui.ListItem(offscreen=True)
item.setProperty('IsPlayable', 'true')
# based on https://github.com/xbmc/xbmc/pull/19530
item.setProperty('ForceResolvePlugin', 'true')
video = {'aired': x.air,
             'year': x.year,
             'episode': int(x.epnumber),
             'sortepisode': int(x.epnumber),
             'plot': spoiler_control_plot(x.summary, spoiler_hide_this),
             'title': title,
             'originaltitle': title,
             'sorttitle': sort_title,
             'tvshowtitle': series_title,
             'mediatype': 'episode',
             'season': 1,
             'sortseason': 1,
             'lastplayed': str(x.view_date) + ' 00:00:00',
             'playcount': 0,
             'overlay': 4
             }
item.setInfo('video', video)
item.setIsFolder(False)
item.setPath(url_for_player)  # url_for_player = the url from API ex. http://10.1.1.100:8111/Stream/123/False/file.mkv
xbmcplugin.setResolvedUrl(handle, True, item)
```

I even tried to add and end directory because there is a cache option there, but nothing changes:
```
xbmcplugin.addDirectoryItem(handle=sys.argv[1], listitem=item, isFolder=False, totalItems=1)
xbmcplugin.endOfDirectory(handle=int(sys.argv[1]), succeeded=True, cacheToDisc=False)
```

**Comment by arnova** [comment]
I also doubt if this is the correct solution. Caching an empty directory is a completely valid use case. It looks like the problem is that we try to fetch a directory from an http directory/server which doesn't have indexing (file listing) enabled. So the question is: how can we properly handle this special case?

**Comment by bigretromike** [comment]
Looks like you are on point with this. I tested with example plugin serving same file from same server, and results in same issue - so thats exclude my video plugin. 

So maybe if its empty and its http/https stream check ?

**Comment by arnova** [comment]
I'm wondering: When http indexing is disabled, shouldn't the GetDirectory-call return failure (false) ? Perhaps this is what needs fixing...

**Comment by bigretromike** [comment]
> I'm wondering: When http indexing is disabled, shouldn't the GetDirectory-call return failure (false) ? Perhaps this is what needs fixing...

Probably but how to do that ?

**Comment by arnova** [comment]
> > I'm wondering: When http indexing is disabled, shouldn't the GetDirectory-call return failure (false) ? Perhaps this is what needs fixing...
> 
> Probably but how to do that ?

I'd investigate why the Open-call in CHTTPDirectory::GetDirectory() seems to return true. Did you try to open the affected url with Curl and see what it does?

**Comment by bigretromike** [comment]
> > > I'm wondering: When http indexing is disabled, shouldn't the GetDirectory-call return failure (false) ? Perhaps this is what needs fixing...
> > 
> > 
> > Probably but how to do that ?
> 
> I'd investigate why the Open-call in CHTTPDirectory::GetDirectory() seems to return true. Did you try to open the affected url with Curl and see what it does?

Both:
curl http://10.1.1.100:8111/Stream/1614/1/False/file.mkv --output -
curl http://10.1.1.100:8111/Stream/1614/1/False --output -
return binary content of .mkv file

**Comment by arnova** [comment]
And what does "curl http://10.1.1.100:8111/Stream/1614/1/False/" give? (note the trailing backslash)

**Comment by bigretromike** [comment]
> And what does "curl http://10.1.1.100:8111/Stream/1614/1/False/" give? (note the trailing backslash)
```
curl -I http://10.1.1.100:8111/Stream/1614/1/False/ --output -
HTTP/1.1 200 OK
Date: Thu, 02 Jun 2022 11:10:42 GMT
Content-Type: video/x-matroska
Server: Shoko Stream Server 1.0
Content-Length: 321998609
Content-Range: bytes 0-321998608/321998609
Accept-Ranges: bytes
api-deprecated-versions: 1.0
```

What about now ? From what I understand Accept-Ranges is only for download content which should indicate that under the URL there is a binary and not Directory.

**Comment by arnova** [comment]
> > And what does "curl http://10.1.1.100:8111/Stream/1614/1/False/" give? (note the trailing backslash)
> 
> ```
> curl -I http://10.1.1.100:8111/Stream/1614/1/False/ --output -
> HTTP/1.1 200 OK
> Date: Thu, 02 Jun 2022 11:10:42 GMT
> Content-Type: video/x-matroska
> Server: Shoko Stream Server 1.0
> Content-Length: 321998609
> Content-Range: bytes 0-321998608/321998609
> Accept-Ranges: bytes
> api-deprecated-versions: 1.0
> ```
> 
> What about now ? From what I understand Accept-Ranges is only for download content which should indicate that under the URL there is a binary and not Directory.

Yes, nice solution. @wsnipex : What do you think?

**Comment by wsnipex** [comment]
looks sane to me

**Review by enen92 — APPROVED** [review]
Looks sane to me too, thanks for digging deeper - next one I'm pretty sure you'll find gold :)

**Comment by bigretromike** [comment]
> Looks sane to me too, thanks for digging deeper - next one I'm pretty sure you'll find gold :)

I pushed a little change because looks like `Accept-Ranges` aint only available when there is binary, this one is more robust. Also any chance to backport it to Matrix ?

**Comment by enen92** [comment]
@bigretromike can you squash the commits and change the pr title?
Also would be nice to know from the list of issues you mention which ones are fixed by your change

**Comment by enen92** [comment]
@bigretromike can you squash the commits and change the pr title?
Also would be nice to know from the list of issues you mention which ones are fixed by your change

**Comment by bigretromike** [comment]
@enen92 will try to squash it (if only I figure out how), also this for sure fix: https://github.com/xbmc/xbmc/issues/21495

**Comment by bigretromike** [comment]
I wont try any more to squash it no need to load jenkins with same tests.

**Comment by fuzzard** [comment]
I've put a no Jenkins tag on it. Get it all cleaned up and we can trigger it after

**Comment by arnova** [comment]
> > Looks sane to me too, thanks for digging deeper - next one I'm pretty sure you'll find gold :)
> 
> I pushed a little change because looks like `Accept-Ranges` aint only available when there is binary, this one is more robust. Also any chance to backport it to Matrix ?

Honestly it seems to me using "Content-Type" seems more fragile since it only covers (some?) music/video types now. Why not  simply check for a Content-Length of Accept-Ranges header?

**Comment by arnova** [comment]
> I wont try any more to squash it no need to load jenkins with same tests.

To squash do:
- "git rebase -i HEAD~9"
- in the editor change "pick" to "fixup" for all commits except the top commit
- "git push -f" to publish to changes to github

**Comment by bigretromike** [comment]
> > I wont try any more to squash it no need to load jenkins with same tests.
> 
> To squash do:
> 
>     * "git rebase -i HEAD~9"
> 
>     * in the editor change "pick" to "fixup" for all commits except the top commit
> 
>     * "git push -f" to publish to changes to github

When I did that I have not only mine commits. I pick mine with fixup but it end up with even more commits :/

**Comment by wsnipex** [comment]
assuming your own repo/fork is origin and the xbmc repo is upstream:
git fetch upstream
git rebase -i upstream/master
git push -f

File: xbmc/filesystem/HTTPDirectory.cpp [hunk 1]
```diff
@@ -75,6 +75,10 @@ bool CHTTPDirectory::GetDirectory(const CURL& url, CFileItemList &items)
   std::string strBuffer;
   if (http.ReadData(strBuffer) && strBuffer.length() > 0)
   {
+    /* if Content-Lenght is found url is pointing to binary so don't treat url as httpdirectory */
```
Comment (arnova) [inline xbmc/filesystem/HTTPDirectory.cpp:78]: binary -> a file

Comment (bigretromike) [reply xbmc/filesystem/HTTPDirectory.cpp]: fixed;

**Comment by arnova** [comment]
Unfortunately rebase/squash didn't work out properly...

File: xbmc/filesystem/HTTPDirectory.cpp [hunk 2]
```diff
@@ -75,6 +75,10 @@ bool CHTTPDirectory::GetDirectory(const CURL& url, CFileItemList &items)
   std::string strBuffer;
   if (http.ReadData(strBuffer) && strBuffer.length() > 0)
   {
+    /* if Content-Lenght is found url is pointing to file so don't treat url as httpdirectory */
```
Comment (arnova) [inline xbmc/filesystem/HTTPDirectory.cpp:78]: One more typo: lenght - > length

Comment (bigretromike) [reply xbmc/filesystem/HTTPDirectory.cpp]: fixed, squashed

**Comment by enen92** [comment]
jenkins build this please

File: xbmc/filesystem/HTTPDirectory.cpp [hunk 3]
```diff
@@ -75,6 +75,10 @@ bool CHTTPDirectory::GetDirectory(const CURL& url, CFileItemList &items)
   std::string strBuffer;
   if (http.ReadData(strBuffer) && strBuffer.length() > 0)
   {
+    /* if Content-Length is found, URL is pointing to file so don't treat URL as HTTPDirectory */
+    if (!StringUtils::EqualsNoCase(http.GetHttpHeader().GetValue("Content-Length"), ""))
```
Comment (enen92) [inline xbmc/filesystem/HTTPDirectory.cpp:79]:
```suggestion
    if (!http.GetHttpHeader().GetValue("Content-Length").empty())
```

**Review by arnova — APPROVED** [review]

**Comment by wsnipex** [comment]
this breaks a bunch of tests: TestHTTPDirectory

**Comment by bigretromike** [comment]
> this breaks a bunch of tests: TestHTTPDirectory

Saw that too, but I don't know why, on my local machine I have all Passes:
```
1>        Start 541: TestHTTPDirectory.IsStarted
1>541/580 Test #541: TestHTTPDirectory.IsStarted ............................................   Passed    0.49 sec
1>        Start 542: TestHTTPDirectory.ApacheDefaultIndex
1>542/580 Test #542: TestHTTPDirectory.ApacheDefaultIndex ...................................   Passed    0.55 sec
1>        Start 543: TestHTTPDirectory.ApacheFancyIndex
1>543/580 Test #543: TestHTTPDirectory.ApacheFancyIndex .....................................   Passed    0.55 sec
1>        Start 544: TestHTTPDirectory.ApacheHtmlIndex
1>544/580 Test #544: TestHTTPDirectory.ApacheHtmlIndex ......................................   Passed    0.55 sec
1>        Start 545: TestHTTPDirectory.BasicIndex
1>545/580 Test #545: TestHTTPDirectory.BasicIndex ...........................................   Passed    0.57 sec
1>        Start 546: TestHTTPDirectory.BasicMultilineIndex
1>546/580 Test #546: TestHTTPDirectory.BasicMultilineIndex ..................................   Passed    0.51 sec
1>        Start 547: TestHTTPDirectory.LighttpDefaultIndex
1>547/580 Test #547: TestHTTPDirectory.LighttpDefaultIndex ..................................   Passed    0.56 sec
1>        Start 548: TestHTTPDirectory.NginxDefaultIndex
1>548/580 Test #548: TestHTTPDirectory.NginxDefaultIndex ....................................   Passed    0.54 sec
```

**Comment by wsnipex** [comment]
I can reproduce here. Content-Length is set and not empty.

does this work for you? Fixes the tests here:

```
diff --git a/xbmc/filesystem/HTTPDirectory.cpp b/xbmc/filesystem/HTTPDirectory.cpp
index 10b6e0b893..1dc7ad1e16 100644
--- a/xbmc/filesystem/HTTPDirectory.cpp
+++ b/xbmc/filesystem/HTTPDirectory.cpp
@@ -76,7 +76,8 @@ bool CHTTPDirectory::GetDirectory(const CURL& url, CFileItemList &items)
   if (http.ReadData(strBuffer) && strBuffer.length() > 0)
   {
     /* if Content-Length is found, URL is pointing to file so don't treat URL as HTTPDirectory */
-    if (!http.GetHttpHeader().GetValue("Content-Length").empty())
+    if (!http.GetHttpHeader().GetValue("Content-Length").empty() &&
+        http.GetHttpHeader().GetValue("Content-type") != "text/html")
       return false;
 
     std::string fileCharset(http.GetProperty(XFILE::FILE_PROPERTY_CONTENT_CHARSET));

```

I'm not sure if this is robust enough

**Comment by bigretromike** [comment]
> I can reproduce here. Content-Length is set and not empty.
> 
> does this work for you? Fixes the tests here:
> 
> ```
> diff --git a/xbmc/filesystem/HTTPDirectory.cpp b/xbmc/filesystem/HTTPDirectory.cpp
> index 10b6e0b893..1dc7ad1e16 100644
> --- a/xbmc/filesystem/HTTPDirectory.cpp
> +++ b/xbmc/filesystem/HTTPDirectory.cpp
> @@ -76,7 +76,8 @@ bool CHTTPDirectory::GetDirectory(const CURL& url, CFileItemList &items)
>    if (http.ReadData(strBuffer) && strBuffer.length() > 0)
>    {
>      /* if Content-Length is found, URL is pointing to file so don't treat URL as HTTPDirectory */
> -    if (!http.GetHttpHeader().GetValue("Content-Length").empty())
> +    if (!http.GetHttpHeader().GetValue("Content-Length").empty() &&
> +        http.GetHttpHeader().GetValue("Content-type") != "text/html")
>        return false;
>  
>      std::string fileCharset(http.GetProperty(XFILE::FILE_PROPERTY_CONTENT_CHARSET));
> ```
> 
> I'm not sure if this is robust enough

It does work

**Comment by bigretromike** [comment]
@wsnipex @enen92 now its builds without errors

**Comment by wsnipex** [comment]
@arnova what do you think? It looks like the usual directory indexes all set content-length

**Comment by arnova** [comment]
> @arnova what do you think? It looks like the usual directory indexes all set content-length

Oh strange, I didn't notice it. It probably depends on the server used. I only tested nginx. In that case your alternative method seems better.

**Comment by bigretromike** [comment]
@wsnipex @arnova do I need to close PR to get it merge or it will eventually get merge by itself ?

**Comment by bigretromike** [comment]
Don't know if you guys know this, but jenkis build it without issue, then after reopeining without code changes couldn't build macos version. If this is known issue then ok. Just want to give you info upfront.

**Comment by enen92** [comment]
@bigretromike osx and ios slaves have issues currently, so don't worry with that. Since the PR has already 2 approvals we'll merge once the building issues are sorted out.

**Comment by wsnipex** [comment]
jenkins build this please

**Pull request merged**